### PR TITLE
ast, optimizer: add flag attribute to expression.

### PR DESCRIPTION
**Pull request by coocood**
We care not only about what the expression is, but also what kind of expression it contains.
Adding flag attribute to expression makes it much easy to determine how to deal with the expression.

The old implementation has `IsStatic` method but that's not enough for our use, some expression is not static but can be evaluated before execution. With flag, `IsStatic` method can be removed.

### CODE REVIEW

File: ast/ast.go [hunk 1]
```diff
@@ -38,13 +38,22 @@ type Node interface {
 	SetText(text string)
 }
 
+// Flags indicates whether an expression contains certain types of expression.
+const (
+	FlagConstant       uint64 = 0
```
Comment (siddontang) [inline ast/ast.go]: is it safe to use 0 for constant flag?

Comment (coocood) [inline ast/ast.go]: 0 means it contains no special expressions, so it is constant.

Comment (shenli) [inline ast/ast.go]: FlagHasNone Is a better name?

File: ast/flag.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+const preEvaluable = FlagHasParamMarker | FlagHasFunc | FlagHasVariable
+
+// IsPreEvaluable checks if the expression can be evaluated before execution.
+func IsPreEvaluable(expr ExprNode) bool {
+	return expr.GetFlag()|preEvaluable == preEvaluable
```
Comment (shenli) [inline ast/flag.go]: return expr.GetFlag()&preEvaluable == preEvaluable?

File: ast/ast.go (hunk 1 above)
Comment (coocood) [inline ast/ast.go]: No, an expression has constant value should not be flagged `HasNone`.

File: ast/flag.go (hunk 2 above)
Comment (coocood) [inline ast/flag.go]: `FlagConstant` is not equal to `preEvaluable`, but it is pre-evaluable.

Comment (siddontang) [inline ast/flag.go]: maybe expr.GetFlag() | preEvaluable > 0 

If expr.GetFlag has another flag, expr.GetFlag()|preEvaluable == preEvaluable is false.

File: ast/flag.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+		x.SetFlag(x.L.GetFlag() | x.R.GetFlag())
+	case *CaseExpr:
+		f.caseExpr(x)
+	case *SubqueryExpr:
+		x.SetFlag(FlagHasSubquery)
+	case *CompareSubqueryExpr:
+		x.SetFlag(x.L.GetFlag() | x.R.GetFlag())
+	case *ColumnNameExpr:
+		x.SetFlag(FlagHasReference)
+	case *DefaultExpr:
```
Comment (siddontang) [inline ast/flag.go]: no need to handle Default?
The default value may be dynamic like timestamp type.

File: ast/flag.go (hunk 2 above)
Comment (coocood) [inline ast/flag.go]: That's what we need it to be.

File: ast/flag_test.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+type testFlagSuite struct {
+}
+
+func (ts *testFlagSuite) TestFlag(c *C) {
+	cases := []struct {
+		expr string
+		flag uint64
+	}{
+		{
+			expr: "1 between 0 and 2",
```
Comment (siddontang) [inline ast/flag_test.go]: no need to write fields here, {"1 between 0 and 2",ast.FlagConstant } is ok.

File: ast/flag.go (hunk 2 above)
Comment (coocood) [inline ast/flag.go]: `expr.GetFlag() | preEvaluable > 0` will always be true.

**Comment by siddontang** [comment]
Rest LGTM

File: ast/flag.go (hunk 2 above)
Comment (siddontang) [inline ast/flag.go]: my fault, expr.GetFlag() & preEvaluable > 0

Comment (coocood) [inline ast/flag.go]: This will test if the expression has those flags, but what we want is to test the expression don't have other flags.

**Comment by shenli** [comment]
LGTM

**Pull request merged**